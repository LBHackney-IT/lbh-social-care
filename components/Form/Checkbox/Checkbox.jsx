import cx from 'classnames';
import PropTypes from 'prop-types';

import ErrorMessage from 'components/ErrorMessage/ErrorMessage';

const Checkbox = ({ label, name, register, required, error, rules }) => (
  <div
    className={cx('govuk-form-group', {
      'govuk-form-group--error': error,
    })}
  >
    <div className="govuk-checkboxes">
      <div className="govuk-checkboxes__item">
        <input
          className={cx('govuk-checkboxes__input', {
            'govuk-input--error': error,
          })}
          id={name}
          name={name}
          type="checkbox"
          ref={rules ? register?.(rules) : register}
        />
        <label className="govuk-label govuk-checkboxes__label" htmlFor={name}>
          {label} {required && <span className="govuk-required">*</span>}
        </label>
      </div>
    </div>
    {error && (
      <ErrorMessage label={error.message} className="govuk-!-margin-top-3" />
    )}
  </div>
);

Checkbox.propTypes = {
  rules: PropTypes.shape(),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  register: PropTypes.func,
  error: PropTypes.shape({ message: PropTypes.string.isRequired }),
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
};

export default Checkbox;
