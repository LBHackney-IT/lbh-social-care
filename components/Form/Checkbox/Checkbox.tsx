import cx from 'classnames';

import ErrorMessage from 'components/ErrorMessage/ErrorMessage';

import type { Checkbox as Props, Option } from 'components/Form/types';

interface Tickbox extends Props {
  value?: string | number;
}

const Tickbox = ({
  error,
  rules,
  register,
  required,
  name,
  label,
  value,
}: Tickbox): React.ReactElement => {
  const id = value ? `${name}-${value}` : name;
  return (
    <div className="govuk-checkboxes__item">
      <input
        className={cx('lbh-checkboxes govuk-checkboxes__input', {
          'govuk-input--error': error,
        })}
        id={id}
        name={name}
        type="checkbox"
        value={value}
        ref={rules ? register?.(rules) : register}
      />
      <label className="govuk-label govuk-checkboxes__label" htmlFor={id}>
        {label} {required && <span className="govuk-required">*</span>}
      </label>
    </div>
  );
};

const CheckBoxWrapper = ({
  children,
  isFieldset,
  ariaDescribedby,
}: {
  children: React.ReactElement;
  isFieldset: boolean;
  ariaDescribedby?: string;
}): React.ReactElement =>
  isFieldset ? (
    <fieldset className="govuk-fieldset" aria-describedby={ariaDescribedby}>
      {children}
    </fieldset>
  ) : (
    children
  );

const Checkbox = ({
  label,
  options,
  name,
  hint,
  error,
  required,
  labelSize = 'm',
  ...otherProps
}: Props): React.ReactElement => {
  return (
    <div
      className={cx('lbh-form-group govuk-form-group', {
        'govuk-form-group--error': error,
      })}
    >
      <CheckBoxWrapper
        isFieldset={Boolean(options)}
        ariaDescribedby={hint && `${name}-hint`}
      >
        <>
          {options && (
            <legend
              className={`govuk-fieldset__legend govuk-fieldset__legend--${labelSize}`}
            >
              {label} {required && <span className="govuk-required">*</span>}
            </legend>
          )}
          {hint && (
            <div id={`${name}-hint`} className="govuk-hint">
              {hint}
            </div>
          )}
          <div className="govuk-checkboxes lbh-checkboxes">
            {options ? (
              options.map((option: Option) => {
                const { value, text } =
                  typeof option === 'string'
                    ? { value: option, text: option }
                    : option;
                return (
                  <Tickbox
                    key={value}
                    {...otherProps}
                    name={name}
                    value={value}
                    label={text}
                  />
                );
              })
            ) : (
              <Tickbox
                {...otherProps}
                name={name}
                label={label}
                required={required}
              />
            )}
          </div>
          {error && (
            <ErrorMessage
              label={error.message}
              className="govuk-!-margin-top-3"
            />
          )}
        </>
      </CheckBoxWrapper>
    </div>
  );
};

export default Checkbox;
