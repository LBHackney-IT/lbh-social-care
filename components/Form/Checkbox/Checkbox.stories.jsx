import Checkbox from './Checkbox';

export default {
  title: 'Form/Checkbox',
  component: Checkbox,
};

const Template = (args) => <Checkbox {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Checkbox',
};

export const withError = Template.bind({});
withError.args = {
  label: 'Checkbox',
  error: { message: 'Ops! There was an error!' },
};
