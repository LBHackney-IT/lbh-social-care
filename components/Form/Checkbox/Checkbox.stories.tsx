import { Story } from '@storybook/react';

import Checkbox from './Checkbox';
import type { Checkbox as Args } from 'components/Form/types';

export default {
  title: 'Form Components/Checkbox',
  component: Checkbox,
};

const Template: Story<Args> = (args) => <Checkbox {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Checkbox',
};

export const multiSelect = Template.bind({});
multiSelect.args = {
  label: 'Multi Checkbox',
  hint: 'You can select multiple values',
  options: ['foo', 'bar', 'foobar'],
};

export const withError = Template.bind({});
withError.args = {
  label: 'Checkbox',
  error: { message: 'Ops! There was an error!' },
};
