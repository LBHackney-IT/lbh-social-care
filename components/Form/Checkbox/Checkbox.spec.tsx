import { render } from '@testing-library/react';

import Checkbox from './Checkbox';

describe('Checkbox', () => {
  const props = {
    name: 'my-checkbox',
    label: 'My Checkbox',
    register: jest.fn(),
  };

  it('renders a checkbox', () => {
    const { getByLabelText } = render(<Checkbox {...props} />);
    const checkbox = getByLabelText(props.label) as HTMLInputElement;
    expect(checkbox).toBeInTheDocument();
    expect(checkbox.id).toEqual(props.name);
    expect(checkbox.name).toEqual(props.name);
  });

  it('should render properly - as Tickbox', () => {
    const { asFragment } = render(<Checkbox {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render properly - as Checkbox', () => {
    const options = [
      {
        text: 'Foo',
        value: 'foo',
      },
      {
        text: 'Bar',
        value: 'bar',
      },
    ];
    const { asFragment } = render(<Checkbox {...props} options={options} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
