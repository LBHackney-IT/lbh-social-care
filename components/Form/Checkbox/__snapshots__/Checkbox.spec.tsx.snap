// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render properly - as Checkbox 1`] = `
<DocumentFragment>
  <div
    class="lbh-form-group govuk-form-group"
  >
    <fieldset
      class="govuk-fieldset"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--m"
      >
        My Checkbox 
      </legend>
      <div
        class="govuk-checkboxes lbh-checkboxes"
      >
        <div
          class="govuk-checkboxes__item"
        >
          <input
            class="lbh-checkboxes govuk-checkboxes__input"
            id="my-checkbox-foo"
            name="my-checkbox"
            type="checkbox"
            value="foo"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="my-checkbox-foo"
          >
            Foo 
          </label>
        </div>
        <div
          class="govuk-checkboxes__item"
        >
          <input
            class="lbh-checkboxes govuk-checkboxes__input"
            id="my-checkbox-bar"
            name="my-checkbox"
            type="checkbox"
            value="bar"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="my-checkbox-bar"
          >
            Bar 
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`Checkbox should render properly - as Tickbox 1`] = `
<DocumentFragment>
  <div
    class="lbh-form-group govuk-form-group"
  >
    <div
      class="govuk-checkboxes lbh-checkboxes"
    >
      <div
        class="govuk-checkboxes__item"
      >
        <input
          class="lbh-checkboxes govuk-checkboxes__input"
          id="my-checkbox"
          name="my-checkbox"
          type="checkbox"
          value=""
        />
        <label
          class="govuk-label govuk-checkboxes__label"
          for="my-checkbox"
        >
          My Checkbox 
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
